docs/spec/relayer/ICS18Environment.tla

```diff
@@ -0,0 +1,314 @@
+
+Heights == 1..MaxHeight \* set of possible heights of the chains in the system
+
+(***************************************************************************
+ Instances of Relayer and Chain
+ ***************************************************************************)
+
+\* We suppose there are two correct relayers in the system, Relayer1 and Relayer2
+\* Relayer1 -- Instance of Relayer.tla
+Relayer1 == INSTANCE Relayer
+            WITH GenerateClientDatagrams <- ClientDatagramsRelayer1,
+                 GenerateConnectionDatagrams <- ConnectionDatagramsRelayer1,
+                 GenerateChannelDatagrams <- ChannelDatagramsRelayer1,
+                 Heights <- Heights,
+                 chainAstore <- chainAstore,
+                 chainBstore <- chainBstore,
+                 relayerHeights <- relayer1Heights
+                 
+\* Relayer2 -- Instance of Relayer.tla      
+Relayer2 == INSTANCE Relayer
+            WITH GenerateClientDatagrams <- ClientDatagramsRelayer2,
+                 GenerateConnectionDatagrams <- ConnectionDatagramsRelayer2,
+                 GenerateChannelDatagrams <- ChannelDatagramsRelayer2,
+                 Heights <- Heights,
+                 chainAstore <- chainAstore,
+                 chainBstore <- chainBstore,
+                 relayerHeights <- relayer2Heights
+
+\* We suppose there are two chains that communicate, ChainA and ChainB
+\* ChainA -- Instance of Chain.tla
+ChainA == INSTANCE Chain
+          WITH ChainID <- "chainA",
+               Heights <- Heights,
+               chainStore <- chainAstore,
+               incomingDatagrams <- incomingDatagramsChainA
+
+\* ChainB -- Instance of Chain.tla 
+ChainB == INSTANCE Chain
+          WITH ChainID <- "chainB",
+               Heights <- Heights,
+               chainStore <- chainBstore,
+               incomingDatagrams <- incomingDatagramsChainB
+
+(***************************************************************************
+ Component actions
+ ***************************************************************************)
+
+\* RelayerAction: either correct relayer takes a step, leaving the other 
+\* variables unchanged 
+RelayerAction ==
+    \/ /\ Relayer1!Next
+       /\ UNCHANGED chainAvars
+       /\ UNCHANGED chainBvars
+       /\ UNCHANGED relayer2Heights
+    \/ /\ Relayer2!Next  
+       /\ UNCHANGED chainAvars
+       /\ UNCHANGED chainBvars
+       /\ UNCHANGED relayer1Heights 
+
+\* ChainAction: either chain takes a step, leaving the other 
+\* variables unchanged       
+ChainAction ==
+    \/ /\ ChainA!Next
+       /\ UNCHANGED chainBvars
+       /\ UNCHANGED relayerVars
+    \/ /\ ChainB!Next  
+       /\ UNCHANGED chainAvars
+       /\ UNCHANGED relayerVars  
+
+(***************************************************************************
+ ICS18Environment actions
+ ***************************************************************************)
+\* Submit datagrams from relayers to chains
+SubmitDatagrams ==
```

Comment:
I guess the idea with this action is to separate datagram computation from submission so we enable more concurrency between relayers?

Reply:
Yes. Once there is a faulty relayer, which would just add datagrams non-deterministically to `outgoingDatagrams`, the environment action `SubmitDatagrams` would also deliver datagrams from the faulty relayer.